src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+                     :do (if (constant-node-p propagated-value-node)
+                             (push (cons var propagated-value-node)
+                                   new-constant-bindings)
+                             (push (cons var propagated-value-node)
+                                   nonconstant-bindings)))
+               (let ((*constant-bindings* (append new-constant-bindings *constant-bindings*)))
+                 (if nonconstant-bindings
+                     (make-node-let
+                      :type (node-type node)
+                      :bindings nonconstant-bindings
+                      :subexpr (funcall *traverse* (node-let-subexpr node) call-stack))
+                     (funcall *traverse* (node-let-subexpr node) call-stack)))))
+
+           (propagate-constants-node-lisp (node call-stack)
+             (declare (type node-lisp node)
+                      (ignore call-stack))
+             (let* ((new-lisp-vars nil)
+                    (let-bindings
+                      (loop :for (lisp-var . coalton-var) :in (node-lisp-vars node)
+                            :for constant-value := (constant-var-value coalton-var :no-error t)
+                            :if constant-value
+                              :collect (let ((new-coalton-var (gentemp (symbol-name coalton-var))))
+                                         (push (cons lisp-var new-coalton-var) new-lisp-vars)
+                                         (cons new-coalton-var constant-value))
+                            :else
+                              :do (push (cons lisp-var coalton-var) new-lisp-vars))))
+               (if let-bindings
+                   (make-node-let
+                    :type (node-type node)
+                    :bindings let-bindings
+                    :subexpr (make-node-lisp
+                              :type (node-type node)
+                              :vars new-lisp-vars
+                              :form (node-lisp-form node)))
+                   (make-node-lisp
+                    :type (node-type node)
+                    :vars new-lisp-vars
+                    :form (node-lisp-form node)))))
+
+           (direct-application-better-infer-types (node call-stack)
+             ;; FIXME: Can we do something similar for non-direct APPLICATION?
+             (let ((constant-propagated-rands
+                     (mapcar (lambda (rand)
+                               (funcall *traverse* rand call-stack))
+                             (node-direct-application-rands node))))
+               (make-node-direct-application
+                :type (node-type node)
+                :rator-type (tc:make-function-type*
+                             (mapcar #'node-type constant-propagated-rands)
+                             (tc:function-return-type (node-type node)))
+                :rator (node-direct-application-rator node)
+                :rands constant-propagated-rands))))
```

Comment:
Since this is being called `:after` the default traversal for `node-direct-application`, the `node` argument to this function `direct-application-better-infer-types` will already have had the default traversal behavior (constant propagation to sub-nodes using the same args (`constant-bindings`) as this node). So `(funcall *traverse*...)` is unnecessary here:
```suggestion




               (make-node-direct-application
                :type (node-type node)
                :rator-type (tc:make-function-type*
                             (mapcar #'node-type (node-direct-application-rands node))
                             (tc:function-return-type (node-type node)))
                :rator (node-direct-application-rator node)
                :rands (node-direct-application-rands node))))
```

Actually, maybe it would make sense to just change the default traversal action on `node-direct-applications` to do exactly what you have here (just with arbitrary `args`) so that the `rator-type` is always updated when the `rands`' types are updated?

Reply:
So, do you mean

1. Keep the `direct-application-better-infer-types` unchanged
2. Use the `:traverse` action instead of `:after` action

Reply:
I mean either

1. Implement the change in the diff and nothing else needs to be changed (`:after` is good as long as you aren't trying to `(funcall *traverse*...)` inside the function to be performed `:after` the main recursive step.)

OR

2. delete the function defined here entirely, and change the action in `make-default-actions` from `src/codegen/traverse.lisp` from
```
    (action (:traverse node-direct-application node &rest args)
      (make-node-direct-application
       :type (node-type node)
       :rator-type (node-direct-application-rator-type node)
       :rator (node-direct-application-rator node)
       :rands (mapcar
               (lambda (node)
                 (apply *traverse* node args))
               (node-direct-application-rands node))))
```
to
```
    (action (:traverse node-direct-application node &rest args)
      (let ((traversed-rands
              (mapcar (lambda (rand)
                        (apply *traverse* rand args))
                      (node-direct-application-rands node))))
        (make-node-direct-application
         :type (node-type node)
         :rator-type (tc:make-function-type*
                      (mapcar #'node-type traversed-rands)
                      (tc:function-return-type (node-type node)))
         :rator (node-direct-application-rator node)
         :rands traversed-rands))))
```
(basically the code you wrote here, but general)
(I am more unsure about this. Maybe it's possible that we would often want applications of `traverse` which don't touch the `rator-type` at all, even if the individual rand `node-type`s get changed?)

--
I guess it would be safe to just go with 1 for now, unless someone thinks 2 would be useful for many other types of traversals.

Reply:
I think for this PR, it would make sense to go with option 1. 

Option 2 may be tested out internally against other codebases using coalton, and then, if nothing breaks, it should be okay. I presume all the type checks and inference steps have already taken place at this point of code. So, I don't see option 2 breaking something. If anything, it might just lead to slightly more stringent checks for whatever the traverse function happens to do. But this seems separate from this PR.

Reply:
Implemented option 1 in 2371482.